src/System.Net.Http/src/System/Net/Http/Managed/AuthenticationHelper.Digest.cs

```diff
@@ -251,128 +244,143 @@ private static string ComputeHash(string data, string algorithm)
             }
         }
 
-        public class DigestResponse
+        internal class DigestResponse
         {
-            public readonly Dictionary<string, string> Parameters = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
+            internal readonly Dictionary<string, string> Parameters = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
 
             // Keep track of request values for this response.
-            public string Nonce;
-            public uint NonceCount;
+            internal uint NonceCount;
 
-            public DigestResponse(string challenge)
+            internal DigestResponse(string challenge)
             {
-                Nonce = null;
                 NonceCount = 1;
 
                 Parse(challenge);
             }
 
-            private unsafe string GetNextKey(ref char* p)
+            private string GetNextKey(ReadOnlySpan<char> p, out int parsedIndex)
             {
-                // It is generally cheaper to change a local and then write back to ref at the end
-                // rather than updating the ref on each operation.
-                char* temp = p;
-
-                StringBuilder sb = StringBuilderCache.Acquire();
+                // Set the index.
+                int i = 0;
 
                 // Skip leading whitespace
-                while (*temp == ' ')
+                while (i < p.Length && p[i] == ' ')
                 {
-                    temp++;
+                    i++;
                 }
 
                 // Start parsing key
-                while (*temp != '=')
-                {
-                    // Key cannot have whitespace
-                    if (*temp == ' ')
-                        break;
+                int start = i;
 
-                    sb.Append(*temp);
-                    temp++;
+                // Parse till '=' is encountered marking end of key.
+                // Key should not contain whitespace, hence break if whitespace found.
+                while (i < p.Length && (p[i] != '=' && p[i] != ' '))
```

Comment:
Should we throw/fail if ' ' is found or '=' is not found?  Same with checks below.  It seems like a bad inputs will result in something being parsed and a string returned, even if not valid.

Reply:
Yeah that's ok, if the input is bad, this will return a key string, but the value parsing will fail, and this key will not be added to the dictionary. The string that we are parsing here is the challenge string we receive from the server on the wire. The GetDigestTokenForCredential understands the semantics of the key,value pairs, and if some key is not found, it will return null, so auth will fail. It is not necessary to throw here.

Reply:
To avoid an extra string allocation here for invalid key, I can return null from this function, and break on null keys in the loop in Parse method.